Implement CurveExt for G1Projective

CPerezz opened the pull request:
This monstrosity which makes almost no-sense (Bls12-381 depending on pasta-curves) is the workaround that unblocks Bls12-381 usage in halo2curves.

For more context on the unblock, see: https://github.com/zkcrypto/group/pull/48 && https://github.com/privacy-scaling-explorations/halo2curves/issues/75

Left to do for the future:
- [ ] `hash_to_curve` impl within `CurveExt` impl.

CPerezz committed:
chore: Add missing fn docs for Fp pub methods

CPerezz committed:
tmp_start

CPerezz committed:
find constants & impl Field/PrimeField for Fp

CPerezz committed:
impl PrimeField for Fp

CPerezz committed:
add missing Fp dependency

CPerezz committed:
impl CurveAffine for G1Affine

CPerezz committed:
impl Ord & PartialOrd for Scalar & Fp

CPerezz committed:
Impl CurveExt for G1Projective

CPerezz commented:
This needs a big cleanup and set a lot of features correctly. 
But it at least compiles. Which allows me if the integration with halo2curves is feasible or not.

CPerezz committed:
try endomorphisms with both N=2 & N=3

CPerezz committed:
fix tests

CPerezz commented:
So we're on a weird situation here where we do have all of the tests passing for all the features. But the endomorphism seems to not work although the params have been derived twice ending in the same result.

CPerezz committed:
Set ZETA with N=2

CPerezz committed:
fix: Bring back original ZETA

CPerezz committed:
chore: Cleanup Fp module

CPerezz committed:
chore: Make clippy happy

CPerezz committed:
chore: Allow nededless borrows

On the way the lib is constructed, the compiler is unable to figure out
that the impl of `Add` and `Mul` that needs to be used (as is the
`const` one is the one done in the module.

Hence why the `&` are used in many places to perform operations.
This means, we need to exclude the clippy checks.

CPerezz committed:
fix: Make no_std friendly the lib again

han0110 commented on src/fp.rs:
```diff
@@ -110,6 +140,63 @@ const R3: Fp = Fp([
 ]);
 
+/// Fp(1/2)
+const TWO_INV: Fp = Fp([
+    0xdcff7fffffffd556,
+    0x0f55ffff58a9ffff,
+    0xb39869507b587b12,
+    0xb23ba5c279c2895f,
+    0x258dd3db21a5d66b,
+    0xd0088f51cbff34d,
+]);
```
```suggestion
/// sage: modulus = 0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab
/// sage: hex(((1 / 2) * (2^384)) % modulus)
/// '0x17fbb8571a006596d3916126f2d14ca26e22d1ec31ebb502633cb57c253c276f855000053ab000011804000000015554'
const TWO_INV: Fp = Fp([
    0x1804000000015554,
    0x855000053ab00001,
    0x633cb57c253c276f,
    0x6e22d1ec31ebb502,
    0xd3916126f2d14ca2,
    0x17fbb8571a006596,
]);
```

han0110 commented on src/fp.rs:
```diff
@@ -110,6 +140,63 @@ const R3: Fp = Fp([
+]);
+
+/// Generator of the field.
+const GENERATOR: Fp = Fp([0x0, 0x0, 0x0, 0x0, 0x0, 0x2]);
```
```suggestion
/// sage: modulus = 0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab
/// sage: hex((2 * (2^384)) % modulus)
/// '0x11ebab9dbb81e28c6cf28d7901622c038b256521ed1f9bcb57605e0db0ddbb51b93c0018d6c40005321300000006554f'
const GENERATOR: Fp = Fp([
    0x321300000006554f,
    0xb93c0018d6c40005,
    0x57605e0db0ddbb51,
    0x8b256521ed1f9bcb,
    0x6cf28d7901622c03,
    0x11ebab9dbb81e28c,
]);
```

han0110 commented on src/fp.rs:
```diff
@@ -110,6 +140,63 @@ const R3: Fp = Fp([
+const GENERATOR: Fp = Fp([0x0, 0x0, 0x0, 0x0, 0x0, 0x2]);
+
+/// Primitive root of unity
+const ROOT_OF_UNITY: Fp = Fp([
+    0x8527e005bf87356d,
+    0xd867814cd448c6d3,
+    0xae5a5b133eddd420,
+    0x356f43ccc999120f,
+    0xc4296fb7e96da4d6,
+    0x1938c01908424cd8,
+]);
```
```suggestion
/// sage: modulus = 0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab
/// sage: hex((GF(modulus)(2) ^ ((modulus - 1) >> 1)) * (2^384))
/// '0x40ab3263eff0206ef148d1ea0f4c069eca8f3318332bb7a07e83a49a2e99d6932b7fff2ed47fffd43f5fffffffcaaae'
const ROOT_OF_UNITY: Fp = Fp([
    0x43f5fffffffcaaae,
    0x32b7fff2ed47fffd,
    0x07e83a49a2e99d69,
    0xeca8f3318332bb7a,
    0xef148d1ea0f4c069,
    0x40ab3263eff0206,
]);
```

han0110 commented on src/fp.rs:
```diff
@@ -110,6 +140,63 @@ const R3: Fp = Fp([
+]);
+
+/// Inverse of the primitive root of unity.
+const ROOT_OF_UNITY_INV: Fp = Fp([
+    0xded7a779aa2898c8,
+    0x53f260542172180b,
+    0xf05b58c2034dddd6,
+    0x4a7cdeacf037ae72,
+    0x7aa8223a111aeea8,
+    0x5046e6d0c9ac33e,
+]);
```
```suggestion
/// sage: modulus = 0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab
/// sage: hex((1 / (GF(modulus)(2) ^ ((modulus - 1) >> 1))) * (2^384))
/// '0x40ab3263eff0206ef148d1ea0f4c069eca8f3318332bb7a07e83a49a2e99d6932b7fff2ed47fffd43f5fffffffcaaae'
const ROOT_OF_UNITY_INV: Fp = Fp([
    0x43f5fffffffcaaae,
    0x32b7fff2ed47fffd,
    0x07e83a49a2e99d69,
    0xeca8f3318332bb7a,
    0xef148d1ea0f4c069,
    0x40ab3263eff0206,
]);
```

han0110 commented on src/fp.rs:
```diff
@@ -110,6 +140,63 @@ const R3: Fp = Fp([
+]);
+
+/// DELTA
+const DELTA: Fp = Fp([
+    0xca172967c7a255d7,
+    0xa5d11a97b6038a99,
+    0x19e243df219f0178,
+    0x74bc071476d308b7,
+    0x73a18c6550d55c00,
+    0x1536fbfcdcf7f2f5,
+]);
```
```suggestion
/// DELTA
/// sage: modulus = 0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab
/// sage: hex((GF(modulus)(2) ^ (2 ^ 1)) * (2^384))
/// '0x9d645513d83de7e8ec9733bbf78ab2fb1d37ebee6ba24d7478fe97a6b0a807f53cc0032fc34000aaa270000000cfff3'
const DELTA: Fp = Fp([
    0xaa270000000cfff3,
    0x53cc0032fc34000a,
    0x478fe97a6b0a807f,
    0xb1d37ebee6ba24d7,
    0x8ec9733bbf78ab2f,
    0x9d645513d83de7e,
]);
```

han0110 commented on src/fp.rs:
```diff
@@ -110,6 +140,63 @@ const R3: Fp = Fp([
+]);
+
+/// ZETA
+// sage: modulus = 0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab
+// sage: GF(modulus).primitive_element() ^ ((modulus - 1) // 3)
+// 793479390729215512621379701633421447060886740281060493010456487427281649075476305620758731620350
+// 5f19672fdf76ce51ba69c6076a0f77eaddb3a93be6f89688de17d813620a00022e01fffffffefffe
+const ZETA: Fp = Fp([
+    0x2e01fffffffefffe,
+    0xde17d813620a0002,
+    0xddb3a93be6f89688,
+    0xba69c6076a0f77ea,
+    0x5f19672fdf76ce51,
+    0x0000000000000000,
+]);
```
```suggestion
/// sage: modulus = 0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab
/// sage: hex((1 / (GF(modulus)(2) ^ ((modulus - 1) // 3))) * (2^384))
/// '0x18f020655463874103f97d6e83d050d28eb60ebe01bacb9e587042afd3851b955dab22461fcda5d2cd03c9e48671f071'
const ZETA: Fp = Fp([
    0xcd03c9e48671f071,
    0x5dab22461fcda5d2,
    0x587042afd3851b95,
    0x8eb60ebe01bacb9e,
    0x03f97d6e83d050d2,
    0x18f0206554638741,
]);
```

han0110 commented on src/fp.rs:
```diff
@@ -158,6 +245,110 @@ impl<'a, 'b> Mul<&'b Fp> for &'a Fp {
+    type Repr = ReprFp;
+
+    fn from_repr(r: Self::Repr) -> CtOption<Self> {
+        Self::from_bytes(&r.0)
```
I think `from_bytes` and `to_bytes` are using big-endian, so we might need to port the `from_repr` and `to_repr` impl here as well

han0110 replied:
I think we'd need to perform a montgomery reduction to get canonical representation. Perhaps we could copy-paste from `from_bytes` and `to_bytes`:

```rust
impl PrimeField for Fp {
    type Repr = ReprFp;

    fn from_repr(repr: Self::Repr) -> CtOption<Self> {
        let mut tmp = Fp([0, 0, 0, 0, 0, 0]);

        tmp.0[0] = u64::from_le_bytes(<[u8; 8]>::try_from(&repr.0[0..8]).unwrap());
        tmp.0[1] = u64::from_le_bytes(<[u8; 8]>::try_from(&repr.0[8..16]).unwrap());
        tmp.0[2] = u64::from_le_bytes(<[u8; 8]>::try_from(&repr.0[16..24]).unwrap());
        tmp.0[3] = u64::from_le_bytes(<[u8; 8]>::try_from(&repr.0[24..32]).unwrap());
        tmp.0[4] = u64::from_le_bytes(<[u8; 8]>::try_from(&repr.0[32..40]).unwrap());
        tmp.0[5] = u64::from_le_bytes(<[u8; 8]>::try_from(&repr.0[40..48]).unwrap());

        // Try to subtract the modulus
        let (_, borrow) = sbb(tmp.0[0], MODULUS[0], 0);
        let (_, borrow) = sbb(tmp.0[1], MODULUS[1], borrow);
        let (_, borrow) = sbb(tmp.0[2], MODULUS[2], borrow);
        let (_, borrow) = sbb(tmp.0[3], MODULUS[3], borrow);
        let (_, borrow) = sbb(tmp.0[4], MODULUS[4], borrow);
        let (_, borrow) = sbb(tmp.0[5], MODULUS[5], borrow);

        // If the element is smaller than MODULUS then the
        // subtraction will underflow, producing a borrow value
        // of 0xffff...ffff. Otherwise, it'll be zero.
        let is_some = (borrow as u8) & 1;

        // Convert to Montgomery form by computing
        // (a.R^0 * R^2) / R = a.R
        tmp *= &R2;

        CtOption::new(tmp, Choice::from(is_some))
    }

    fn to_repr(&self) -> Self::Repr {
        // Turn into canonical form by computing
        // (a.R) / R = a
        let tmp = Fp::montgomery_reduce(
            self.0[0], self.0[1], self.0[2], self.0[3], self.0[4], self.0[5], 0, 0, 0, 0, 0, 0,
        );

        let mut res = [0; 48];
        res[0..8].copy_from_slice(&tmp.0[0].to_le_bytes());
        res[8..16].copy_from_slice(&tmp.0[1].to_le_bytes());
        res[16..24].copy_from_slice(&tmp.0[2].to_le_bytes());
        res[24..32].copy_from_slice(&tmp.0[3].to_le_bytes());
        res[32..40].copy_from_slice(&tmp.0[4].to_le_bytes());
        res[40..48].copy_from_slice(&tmp.0[5].to_le_bytes());

        ReprFp(res)
    }

    // ...
}
```

And I'd suggest to add a test for `from_repr` and `to_repr` like:

```rust
#[test]
fn test_repr() {
    for value in [Fp::ZERO, Fp::ONE, Fp::ROOT_OF_UNITY] {
        assert_eq!(Fp::from_repr(value.to_repr()).unwrap(), value);
    }
}
```

Also I think the wrapper `ReprFp` seems a little bit verbose to me so I'd propose to just have `[u8; 48]` as the `Repr`, which is what other fields are using.

kilic replied:
Unfortunately `[u8;48]` does not implement `Default` bound. So this was also the case with extension field representatoins

han0110 replied:
Ah I see it now

CPerezz replied:
So what I've done is basically add LE for `ReprFp` and BE for the `from/to_bytes`. I'll try to document it somewhere and that should be it. We should have the two flavours.

Is this ok?

han0110 reviewed (commented):
Some constants of `Fp` need to be updated since it doesn't have `from_raw` to use, I've generated locally and add some comments.

Also perhaps we could add a test for these constans below like:

```rust
#[test]
fn constants() {
    assert_eq!(Fp::ONE.double(), GENERATOR);
    assert_eq!(Fp::ONE, Fp::ONE.double() * TWO_INV);
    assert_eq!(Fp::ONE, ROOT_OF_UNITY.pow([1 << Fp::S]));
    assert_eq!(Fp::ONE, ROOT_OF_UNITY * ROOT_OF_UNITY_INV);
    // sage: modulus = 0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab
    // sage: hex((modulus - 1) >> 1)
    // '0xd0088f51cbff34d258dd3db21a5d66bb23ba5c279c2895fb39869507b587b120f55ffff58a9ffffdcff7fffffffd555'
    let t = [
        0xdcff7fffffffd555,
        0x0f55ffff58a9ffff,
        0xb39869507b587b12,
        0xb23ba5c279c2895f,
        0x258dd3db21a5d66b,
        0xd0088f51cbff34d,
    ];
    assert_eq!(Fp::ONE, DELTA.pow(t));
    assert_eq!(Fp::ONE, ZETA.pow([3]));
}
```

CPerezz commented on src/fp.rs:
```diff
@@ -158,6 +245,110 @@ impl<'a, 'b> Mul<&'b Fp> for &'a Fp {
+    const CAPACITY: u32 = Self::NUM_BITS - 1;
+    const TWO_INV: Self = TWO_INV;
+    const MULTIPLICATIVE_GENERATOR: Self = GENERATOR;
+    const S: u32 = 32;
```
Ohhh my bad. I missed this from the copy I did from  `Scalar`

CPerezz commented on src/g1.rs:
```diff
@@ -436,6 +456,54 @@ fn endomorphism(p: &G1Affine) -> G1Affine {
+        let p: G1Affine = Self {
+            x,
+            y,
+            infinity: if x == Self::Base::ZERO && y == Self::Base::ONE {
+                Choice::from(1u8)
+            } else {
+                Choice::from(0u8)
+            },
```
Makes sense! Thanks!

CPerezz committed:
Apply suggestions from Han's review

These mostly consist on moving all the `Fp` constants into the `Montgomery form` since `Scalar::from_raw` is missing.

Co-authored-by: Han <tinghan0110@gmail.com>

CPerezz commented:
> Some constants of `Fp` need to be updated since it doesn't have `from_raw` to use, I've generated locally and add some comments.
> 
> Also perhaps we could add a test for these constans below like:
> 
> ```rust
> #[test]
> fn constants() {
>     assert_eq!(Fp::ONE.double(), GENERATOR);
>     assert_eq!(Fp::ONE, Fp::ONE.double() * TWO_INV);
>     assert_eq!(Fp::ONE, ROOT_OF_UNITY.pow([1 << Fp::S]));
>     assert_eq!(Fp::ONE, ROOT_OF_UNITY * ROOT_OF_UNITY_INV);
>     // sage: modulus = 0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab
>     // sage: hex((modulus - 1) >> 1)
>     // '0xd0088f51cbff34d258dd3db21a5d66bb23ba5c279c2895fb39869507b587b120f55ffff58a9ffffdcff7fffffffd555'
>     let t = [
>         0xdcff7fffffffd555,
>         0x0f55ffff58a9ffff,
>         0xb39869507b587b12,
>         0xb23ba5c279c2895f,
>         0x258dd3db21a5d66b,
>         0xd0088f51cbff34d,
>     ];
>     assert_eq!(Fp::ONE, DELTA.pow(t));
>     assert_eq!(Fp::ONE, ZETA.pow([3]));
> }
> ```

I indeed had this kind of test in https://github.com/privacy-scaling-explorations/bls12_381/blob/tmp_curve_ext/src/scalar.rs#L819-L849

And assumed I had it too for `Fp`. Adding it!!

CPerezz committed:
chore: Add tests for Fp constants & nits fix

CPerezz committed:
fix: Use little endian for the `ReprFp` type

CPerezz committed:
fix: Adjust LE usage for ReprFp & add test

han0110 reviewed (approved):
LGTM
